File: invert-binary-tree/invert-binary-tree.c
#include <stdio.h>
#include <assert.h>
#include "../tree_node.h"

struct TreeNode* invertTree(struct TreeNode* root){
    if (root == NULL) {
        return NULL;
    }
    
    invertTree(root->left);
    invertTree(root->right);

    struct TreeNode* tmp = NULL;
    tmp = root->left;
    root->left = root->right;
    root->right = tmp;

    return root;
}

int main() {
    struct TreeNode n0, n1, n2, n3, n4, n5, n6;
    
    init_tree_node(&n1, 2, &n3, &n4);
    init_tree_node(&n3, 1, NULL, NULL);
    init_tree_node(&n4, 3, NULL, NULL);
    
    init_tree_node(&n2, 7, &n5, &n6);
    init_tree_node(&n5, 6, NULL, NULL);
    init_tree_node(&n6, 9, NULL, NULL);

    init_tree_node(&n0, 4, &n1, &n2);

    print_tree(&n0, 0);
    invertTree(&n0);
    print_tree(&n0, 0);

    return 0;
}

File: n150/invert-binary-tree/invert-binary-tree.py
# https://leetcode.com/problems/invert-binary-tree

from TreeNode import TreeNode

class Solution(object):
    def step(self, node):
        if node == None:
            return
        self.step(node.left)
        self.step(node.right)
        node.left, node.right = node.right, node.left

    def invertTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        self.step(root)
        return root
        

if __name__ == "__main__":
    sol = Solution()
    tests = [
        (
            [4,2,7,1,3,6,9],
            [4,7,2,9,6,3,1],
        ),
    ]
    for (nums, solution) in tests:
        print("-----------------------------------")
        tree = TreeNode.from_array(nums)
        print("tree:")
        tree.print_tree()
        solution = TreeNode.from_array(solution)
        solution.print_tree()
        print("solution:")
        result = sol.invertTree(tree)
        print("result:")
        result.print_tree()
    print("✅ All tests passed")

